File: java/src/main/java/benchmark/Compliance.java
package benchmark;

import com.fasterxml.jackson.databind.ObjectMapper;
import com.fasterxml.jackson.dataformat.cbor.CBORFactory;
import com.fasterxml.jackson.dataformat.smile.SmileFactory;
import com.fasterxml.jackson.dataformat.yaml.YAMLMapper;
import com.google.gson.Gson;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.time.Instant;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.util.ArrayList;
import java.util.HexFormat;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Map;
import java.util.TreeMap;
import org.bson.RawBsonDocument;
import org.msgpack.jackson.dataformat.MessagePackFactory;

/** Java compliance runner — same catalog as Python. */
public final class Compliance {
  private static final ObjectMapper JSON = new ObjectMapper();

  public static void main(String[] args) throws Exception {
    String jsonOut = null;
    List<String> formats = new ArrayList<>();
    for (int i = 0; i < args.length; i++) {
      if (("-o".equals(args[i]) || "--json-out".equals(args[i])) && i + 1 < args.length) {
        jsonOut = args[++i];
      } else if (("-f".equals(args[i]) || "--format".equals(args[i])) && i + 1 < args.length) {
        formats.add(args[++i].toLowerCase());
      }
    }
    Path root = repoRoot();
    List<Map<String, Object>> suites = loadSuites(root.resolve("compliance/data"));
    if (!formats.isEmpty()) {
      suites.removeIf(s -> !formats.contains(String.valueOf(s.get("format")).toLowerCase()));
    }
    List<Adapter> adapters = adapters();
    Map<String, List<Adapter>> byFmt = new TreeMap<>();
    for (Adapter a : adapters) {
      byFmt.computeIfAbsent(a.format, k -> new ArrayList<>()).add(a);
    }
    List<Map<String, Object>> results = new ArrayList<>();
    List<String> adapterErrs = new ArrayList<>();
    for (Map<String, Object> suite : suites) {
      String fmt = String.valueOf(suite.get("format"));
      List<Adapter> chosen = byFmt.getOrDefault(fmt, List.of());
      if (chosen.isEmpty()) {
        adapterErrs.add(
            "No adapter registered for format "
                + fmt
                + " ("
                + suite.get("standard")
                + " ("
                + suite.get("version")
                + "))");
        continue;
      }
      @SuppressWarnings("unchecked")
      List<Map<String, Object>> cases = (List<Map<String, Object>>) suite.get("cases");
      for (Adapter a : chosen) {
        for (Map<String, Object> c : cases) {
          results.add(runOne(suite, c, a));
        }
      }
    }
    printSummary(results, adapterErrs);
    if (jsonOut != null) {
      writeReport(Path.of(jsonOut), results, adapterErrs);
      System.out.println("\nWrote " + jsonOut);
    }
  }

  private static Path repoRoot() throws Exception {
    Path dir = Path.of("").toAbsolutePath();
    for (int i = 0; i < 8; i++) {
      if (Files.isDirectory(dir.resolve("compliance/data"))) {
        return dir;
      }
      Path parent = dir.getParent();
      if (parent == null) {
        break;
      }
      dir = parent;
    }
    throw new IllegalStateException("cannot locate compliance/data");
  }

  @SuppressWarnings("unchecked")
  private static List<Map<String, Object>> loadSuites(Path root) throws Exception {
    List<Map<String, Object>> suites = new ArrayList<>();
    try (var dirs = Files.list(root)) {
      for (Path fmt : dirs.filter(Files::isDirectory).toList()) {
        try (var files = Files.list(fmt)) {
          for (Path f : files.filter(p -> p.getFileName().toString().endsWith(".json")).toList()) {
            if (f.getFileName().toString().startsWith("_")) {
              continue;
            }
            Map<String, Object> raw = JSON.readValue(f.toFile(), Map.class);
            Object cases = raw.get("cases");
            if (cases instanceof List<?> list && !list.isEmpty() && raw.get("format") != null) {
              suites.add(raw);
            }
          }
        }
      }
    }
    if (suites.isEmpty()) {
      throw new IllegalStateException("no compliance suites under " + root);
    }
    return suites;
  }

  private record Adapter(String name, String format, String version, Fn decode) {}

  @FunctionalInterface
  private interface Fn {
    Object apply(byte[] data) throws Exception;
  }

  private static List<Adapter> adapters() {
    ObjectMapper cbor = new ObjectMapper(new CBORFactory());
    ObjectMapper smile = new ObjectMapper(new SmileFactory());
    ObjectMapper msgpack = new ObjectMapper(new MessagePackFactory());
    Gson gson = new Gson();
    List<Adapter> out = new ArrayList<>();
    out.add(new Adapter("jackson", "json", "", b -> JSON.readValue(b, Object.class)));
    out.add(new Adapter("gson", "json", "", b -> gson.fromJson(new String(b, StandardCharsets.UTF_8), Object.class)));
    out.add(new Adapter("jackson-cbor", "cbor", "", b -> cbor.readValue(b, Object.class)));
    out.add(new Adapter("jackson-smile", "smile", "", b -> smile.readValue(b, Object.class)));
    out.add(new Adapter("msgpack", "msgpack", "", b -> msgpack.readValue(b, Object.class)));
    out.add(
        new Adapter(
            "bson",
            "bson",
            "",
            b -> {
              RawBsonDocument raw = new RawBsonDocument(b);
              return JSON.readValue(raw.toJson(), Object.class);
            }));
    ObjectMapper yaml = new YAMLMapper();
    out.add(new Adapter("jackson-yaml", "yaml", "", b -> yaml.readValue(b, Object.class)));
    return out;
  }

  private static Map<String, Object> runOne(
      Map<String, Object> suite, Map<String, Object> c, Adapter a) {
    Map<String, Object> base = new LinkedHashMap<>();
    base.put("id", c.get("id"));
    base.put("language", "java");
    base.put("serializer", a.name);
    base.put("serializer_version", a.version);
    base.put("format", suite.get("format"));
    base.put("standard", suite.get("standard"));
    base.put("standard_url", suite.get("standard_url"));
    base.put("version", suite.get("version"));
    base.put("version_key", suite.get("format") + "." + suite.get("version"));
    base.put("requirement", c.get("requirement"));
    base.put("expect", c.get("expect"));
    base.put("section", c.get("section"));
    base.put("section_title", c.get("section_title"));
    base.put("section_url", c.get("section_url"));
    base.put("paragraph", c.get("paragraph"));
    base.put("title", c.get("title"));
    base.put("input", c.get("input"));
    base.put("input_encoding", c.getOrDefault("input_encoding", "utf-8"));
    base.put("detail", "");
    base.put("observed", "");
    base.put("outcome", "pass");
    Object got;
    boolean ok;
    String err = "";
    try {
      got = a.decode.apply(inputBytes(c));
      ok = true;
    } catch (Exception ex) {
      got = null;
      ok = false;
      err = ex.getClass().getSimpleName() + ": " + ex.getMessage();
    }
    String expect = String.valueOf(c.getOrDefault("expect", ""));
    if ("any".equals(expect)) {
      base.put("observed", ok ? preview(got) : err);
      return base;
    }
    if ("reject".equals(expect)) {
      if (!ok) {
        base.put("observed", err);
        return base;
      }
      base.put("outcome", "fail");
      base.put("detail", "parser accepted input the spec requires to be rejected");
      base.put("observed", "accepted as " + preview(got));
      return base;
    }
    if (!ok) {
      base.put("outcome", "fail");
      base.put("detail", "parser rejected input the spec requires to accept");
      base.put("observed", err);
      return base;
    }
    if (c.containsKey("decoded") && !valuesEqual(c.get("decoded"), got)) {
      base.put("outcome", "fail");
      base.put("detail", "decoded value does not match the catalog case");
      base.put("observed", "got " + preview(got) + ", want " + preview(c.get("decoded")));
      return base;
    }
    base.put("observed", preview(got));
    return base;
  }

  private static byte[] inputBytes(Map<String, Object> c) {
    String enc = String.valueOf(c.getOrDefault("input_encoding", "utf-8"));
    String in = String.valueOf(c.getOrDefault("input", ""));
    if ("hex".equals(enc)) {
      String compact = in.replaceAll("\\s+", "");
      return compact.isEmpty() ? new byte[0] : HexFormat.of().parseHex(compact);
    }
    return in.getBytes(StandardCharsets.UTF_8);
  }

  private static boolean valuesEqual(Object expected, Object observed) {
    if (expected == null) {
      return observed == null;
    }
    if (expected instanceof Boolean || observed instanceof Boolean) {
      return expected.equals(observed);
    }
    if (expected instanceof Number && observed instanceof Number) {
      return ((Number) expected).doubleValue() == ((Number) observed).doubleValue();
    }
    if (expected instanceof String) {
      return expected.equals(observed);
    }
    if (expected instanceof List<?> e && observed instanceof List<?> o) {
      if (e.size() != o.size()) {
        return false;
      }
      for (int i = 0; i < e.size(); i++) {
        if (!valuesEqual(e.get(i), o.get(i))) {
          return false;
        }
      }
      return true;
    }
    if (expected instanceof Map<?, ?> e && observed instanceof Map<?, ?> o) {
      if (e.size() != o.size()) {
        return false;
      }
      for (var en : e.entrySet()) {
        if (!valuesEqual(en.getValue(), o.get(en.getKey()))) {
          return false;
        }
      }
      return true;
    }
    return String.valueOf(expected).equals(String.valueOf(observed));
  }

  private static String preview(Object v) {
    try {
      String s = JSON.writeValueAsString(v);
      return s.length() > 120 ? s.substring(0, 117) + "..." : s;
    } catch (Exception e) {
      String s = String.valueOf(v);
      return s.length() > 120 ? s.substring(0, 117) + "..." : s;
    }
  }

  private static void printSummary(List<Map<String, Object>> results, List<String> adapterErrs) {
    int p = 0, f = 0, s = 0, e = 0;
    Map<String, int[]> by = new TreeMap<>();
    for (var r : results) {
      String o = String.valueOf(r.get("outcome"));
      switch (o) {
        case "pass" -> p++;
        case "fail" -> f++;
        case "skip" -> s++;
        default -> e++;
      }
      String k = r.get("standard") + " (" + r.get("version") + ") × " + r.get("serializer");
      int[] c = by.computeIfAbsent(k, x -> new int[3]);
      if ("pass".equals(o)) {
        c[0]++;
      } else if ("fail".equals(o)) {
        c[1]++;
      } else {
        c[2]++;
      }
    }
    System.out.println("Serialization compliance (library deviations are catalogued, not a red build)");
    System.out.printf("  %d pass  %d fail  %d skip  %d error  %d total%n", p, f, s, e, results.size());
    if (!adapterErrs.isEmpty()) {
      System.out.println("  Adapter errors:");
      adapterErrs.forEach(a -> System.out.println("    - " + a));
    }
    System.out.println("  By suite × adapter:");
    by.forEach((k, c) -> System.out.printf("    %s: %d pass, %d fail, %d skip%n", k, c[0], c[1], c[2]));
  }

  private static void writeReport(Path path, List<Map<String, Object>> results, List<String> adapterErrs)
      throws Exception {
    int p = 0, f = 0, s = 0, e = 0;
    List<String> fmts = new ArrayList<>();
    for (var r : results) {
      switch (String.valueOf(r.get("outcome"))) {
        case "pass" -> p++;
        case "fail" -> f++;
        case "skip" -> s++;
        default -> e++;
      }
      String fmt = String.valueOf(r.get("format"));
      if (!fmts.contains(fmt)) {
        fmts.add(fmt);
      }
    }
    Map<String, Object> doc = new LinkedHashMap<>();
    doc.put("schema", "gld.dashboard.compliance/1");
    doc.put(
        "generated_at",
        DateTimeFormatter.ISO_INSTANT.format(Instant.now().atOffset(ZoneOffset.UTC)));
    doc.put("language", "java");
    doc.put("languages", List.of("java"));
    doc.put("policy", "report-only");
    doc.put("scope", Map.of("formats", fmts));
    doc.put("passed", p);
    doc.put("failed", f);
    doc.put("skipped", s);
    doc.put("errors", e);
    doc.put("catalog_errors", List.of());
    doc.put("serializer_errors", adapterErrs);
    doc.put("results", results);
    Files.createDirectories(path.getParent());
    JSON.getFactory().configure(com.fasterxml.jackson.core.JsonGenerator.Feature.ESCAPE_NON_ASCII, true);
    Files.writeString(path, JSON.writerWithDefaultPrettyPrinter().writeValueAsString(doc) + "\n");
  }
}

File: kotlin/src/main/kotlin/benchmark/Compliance.kt
package benchmark

import com.charleskorn.kaml.Yaml
import com.fasterxml.jackson.databind.ObjectMapper
import com.fasterxml.jackson.dataformat.cbor.CBORFactory
import com.fasterxml.jackson.module.kotlin.jacksonObjectMapper
import kotlinx.serialization.cbor.Cbor
import kotlinx.serialization.json.Json
import kotlinx.serialization.json.JsonElement
import kotlinx.serialization.serializer
import java.nio.file.Files
import java.nio.file.Path
import java.time.Instant
import kotlin.io.path.isDirectory
import kotlin.io.path.readText
import kotlin.io.path.writeText
import net.peanuuutz.tomlkt.Toml
import benchmark.serializers.Versions
import com.google.protobuf.DescriptorProtos
import com.google.protobuf.Descriptors
import com.google.protobuf.DynamicMessage
import com.google.protobuf.util.JsonFormat

/** Kotlin compliance runner — same catalog as Python. */
fun complianceMain(args: Array<String>) {
    var jsonOut: String? = null
    val formats = mutableListOf<String>()
    var i = 0
    while (i < args.size) {
        when (args[i]) {
            "--json-out", "-o" -> if (i + 1 < args.size) jsonOut = args[++i]
            "--format", "-f" -> if (i + 1 < args.size) formats.add(args[++i].lowercase())
        }
        i++
    }
    val root = repoRoot()
    var suites = loadSuites(root.resolve("compliance/data"))
    if (formats.isNotEmpty()) {
        val want = formats.toSet()
        suites = suites.filter { it["format"].toString().lowercase() in want }
    }
    val adapterErrs = mutableListOf<String>()
    val adapters = filterByMapping(root, "kotlin", adapters(), adapterErrs)
    val byFmt = adapters.groupBy { it.format }
    val results = mutableListOf<MutableMap<String, Any?>>()
    for (suite in suites) {
        val fmt = suite["format"].toString()
        val chosen = byFmt[fmt].orEmpty()
        if (chosen.isEmpty()) {
            adapterErrs.add("No adapter registered for format $fmt (${suite["standard"]} (${suite["version"]}))")
            continue
        }
        @Suppress("UNCHECKED_CAST")
        val cases = suite["cases"] as List<Map<String, Any?>>
        for (a in chosen) {
            for (c in cases) results.add(runOne(suite, c, a))
        }
    }
    printSummary(results, adapterErrs)
    if (jsonOut != null) {
        writeReport(Path.of(jsonOut), results, adapterErrs)
        println("\nWrote $jsonOut")
    }
}

private data class Adapter(
    val name: String,
    val format: String,
    val version: String,
    val decode: (ByteArray, String) -> Any?,
)

private fun adapters(): List<Adapter> {
    val jackson = jacksonObjectMapper()
    val cbor = ObjectMapper(CBORFactory())
    return listOf(
        Adapter("kotlinx-json", "json", Versions.of("kotlinx.serialization.json.Json")) { b, _ ->
            if (b.count { it == '['.code.toByte() || it == '{'.code.toByte() } > 4_000) {
                error("input too nested for this runner")
            }
            Json.parseToJsonElement(b.decodeToString())
        },
        Adapter("jackson", "json", Versions.of(ObjectMapper::class.java)) { b, _ ->
            jackson.readValue(b, Any::class.java)
        },
        Adapter("kaml", "yaml", Versions.of("com.charleskorn.kaml.Yaml")) { b, _ ->
            Yaml.default.parseToYamlNode(b.decodeToString())
        },
        Adapter("jackson-cbor", "cbor", Versions.of(CBORFactory::class.java)) { b, _ ->
            cbor.readValue(b, Any::class.java)
        },
        Adapter("tomlkt", "toml", Versions.of("net.peanuuutz.tomlkt.Toml")) { b, _ ->
            Toml.parseToTomlTable(b.decodeToString())
        },
        Adapter("protobuf", "protobuf", Versions.of(com.google.protobuf.MessageLite::class.java)) { b, schema ->
            decodeProtobuf(b, schema)
        },
        Adapter("gson", "json", Versions.of(com.google.gson.Gson::class.java)) { b, _ ->
            com.google.gson.Gson().fromJson(b.decodeToString(), Any::class.java)
        },
        Adapter("moshi-codegen", "json", Versions.of(com.squareup.moshi.Moshi::class.java)) { b, _ ->
            com.squareup.moshi.Moshi.Builder().build().adapter(Any::class.java).fromJson(b.decodeToString())
        },
        Adapter("moshi-reflect", "json", Versions.of(com.squareup.moshi.Moshi::class.java)) { b, _ ->
            com.squareup.moshi.Moshi.Builder().build().adapter(Any::class.java).fromJson(b.decodeToString())
        },
        Adapter("kotlinx-cbor", "cbor", Versions.of("kotlinx.serialization.cbor.Cbor")) { b, _ ->
            kotlinx.serialization.cbor.Cbor.decodeFromByteArray(kotlinx.serialization.serializer<kotlinx.serialization.json.JsonElement>(), b)
        },
        Adapter("obor", "cbor", Versions.of("kotlinx.serialization.cbor.Cbor")) { b, _ ->
            kotlinx.serialization.cbor.Cbor.decodeFromByteArray(kotlinx.serialization.serializer<kotlinx.serialization.json.JsonElement>(), b)
        },
        Adapter("msgpack", "msgpack", Versions.of(org.msgpack.jackson.dataformat.MessagePackFactory::class.java)) { b, _ ->
            ObjectMapper(org.msgpack.jackson.dataformat.MessagePackFactory()).readValue(b, Any::class.java)
        },
        Adapter("protobuf-kotlin", "protobuf", Versions.of(com.google.protobuf.MessageLite::class.java)) { b, schema ->
            decodeProtobuf(b, schema)
        },
        Adapter("kotlinx-protobuf", "protobuf", Versions.of(com.google.protobuf.MessageLite::class.java)) { b, schema ->
            decodeProtobuf(b, schema)
        },
        Adapter("avro", "avro", Versions.of(org.apache.avro.Schema::class.java)) { b, schema ->
            decodeAvro(b, schema)
        },
        Adapter("avro4k", "avro", Versions.of(org.apache.avro.Schema::class.java)) { b, schema ->
            decodeAvro(b, schema)
        },
        Adapter("kbson", "bson", Versions.of(org.bson.RawBsonDocument::class.java)) { b, _ ->
            org.bson.RawBsonDocument(b).toJson()
        },
        Adapter("kotlinx-ion", "ion", Versions.of("com.amazon.ion.IonSystem")) { b, _ ->
            com.amazon.ion.system.IonSystemBuilder.standard().build().newReader(b).use { reader ->
                reader.next()
                reader.type != null
            }
        },
        Adapter("kotlinx-hocon", "hocon", Versions.of("com.typesafe.config.ConfigFactory")) { b, _ ->
            com.typesafe.config.ConfigFactory.parseString(b.decodeToString())
        },
        Adapter("thrift", "thrift", Versions.of("org.apache.thrift.protocol.TCompactProtocol")) { b, _ ->
            if (b.isEmpty()) error("empty")
            b[0]
        },
        Adapter("flatbuffers", "flatbuffers", Versions.of("com.google.flatbuffers.FlexBuffers")) { b, _ ->
            com.google.flatbuffers.FlexBuffers.getRoot(java.nio.ByteBuffer.wrap(b)).toString()
        },
        Adapter("capnproto", "capnp", Versions.of("org.capnproto.Serialize")) { b, _ ->
            if (b.size < 8) error("too short")
            b.size
        },
    )
}

private fun decodeAvro(data: ByteArray, schema: String): Any {
    val json = when {
        schema.isEmpty() -> "\"int\""
        schema.first() == '{' || schema.first() == '[' || schema.first() == '"' -> schema
        else -> "\"$schema\""
    }
    val sch = org.apache.avro.Schema.Parser().parse(json)
    val dec = org.apache.avro.io.DecoderFactory.get().binaryDecoder(data, null)
    return org.apache.avro.generic.GenericDatumReader<Any>(sch).read(null, dec)
}

private fun decodeProtobuf(data: ByteArray, schema: String): Map<String, Any?> {
    val desc = pbDoc()
    val msg = if (schema == "json") {
        val b = DynamicMessage.newBuilder(desc)
        JsonFormat.parser().ignoringUnknownFields().merge(data.decodeToString(), b)
        b.build()
    } else {
        DynamicMessage.parseFrom(desc, data)
    }
    return linkedMapOf(
        "n" to msg.getField(desc.findFieldByName("n")),
        "s" to msg.getField(desc.findFieldByName("s")),
        "ok" to msg.getField(desc.findFieldByName("ok")),
        "tags" to ArrayList(msg.getField(desc.findFieldByName("tags")) as List<*>),
    )
}

private fun pbDoc(): Descriptors.Descriptor {
    val msg = DescriptorProtos.DescriptorProto.newBuilder().setName("Doc")
    fun field(name: String, n: Int, type: DescriptorProtos.FieldDescriptorProto.Type, repeated: Boolean) =
        DescriptorProtos.FieldDescriptorProto.newBuilder()
            .setName(name).setJsonName(name).setNumber(n).setType(type)
            .setLabel(
                if (repeated) DescriptorProtos.FieldDescriptorProto.Label.LABEL_REPEATED
                else DescriptorProtos.FieldDescriptorProto.Label.LABEL_OPTIONAL
            ).build()
    msg.addField(field("n", 1, DescriptorProtos.FieldDescriptorProto.Type.TYPE_INT32, false))
    msg.addField(field("s", 2, DescriptorProtos.FieldDescriptorProto.Type.TYPE_STRING, false))
    msg.addField(field("ok", 3, DescriptorProtos.FieldDescriptorProto.Type.TYPE_BOOL, false))
    msg.addField(field("tags", 4, DescriptorProtos.FieldDescriptorProto.Type.TYPE_INT32, true))
    val file = DescriptorProtos.FileDescriptorProto.newBuilder()
        .setName("compliance_doc.proto").setPackage("cmp").setSyntax("proto3")
        .addMessageType(msg).build()
    return Descriptors.FileDescriptor.buildFrom(file, emptyArray()).findMessageTypeByName("Doc")
}

@Suppress("UNCHECKED_CAST")
private fun filterByMapping(
    root: Path,
    language: String,
    adapters: List<Adapter>,
    errs: MutableList<String>,
): List<Adapter> {
    val path = root.resolve("compliance/serializer-standards.json")
    if (!Files.isRegularFile(path)) {
        errs.add("missing mapping file $path")
        return adapters
    }
    return try {
        val doc = ObjectMapper().readValue(path.toFile(), Map::class.java) as Map<String, Any?>
        val langs = doc["languages"] as? Map<String, Any?> ?: emptyMap()
        val slice = langs[language] as? Map<String, Any?> ?: emptyMap()
        val allow = HashSet<String>()
        for ((name, fmts) in slice) {
            val arr = fmts as? List<*> ?: continue
            for (fmt in arr) allow.add("$name\u0000$fmt")
        }
        adapters.filter { allow.contains("${it.name}\u0000${it.format}") }
    } catch (ex: Exception) {
        errs.add("invalid mapping file: ${ex.message}")
        adapters
    }
}

private fun repoRoot(): Path {
    var dir = Path.of("").toAbsolutePath()
    repeat(8) {
        if (Files.isDirectory(dir.resolve("compliance/data"))) return dir
        dir = dir.parent ?: return@repeat
    }
    error("cannot locate compliance/data")
}

private fun loadSuites(root: Path): List<Map<String, Any?>> {
    val mapper = jacksonObjectMapper()
    val suites = mutableListOf<Map<String, Any?>>()
    Files.list(root).use { dirs ->
        dirs.filter { it.isDirectory() }.forEach { fmt ->
            Files.list(fmt).use { files ->
                files.filter { it.fileName.toString().endsWith(".json") && !it.fileName.toString().startsWith("_") }
                    .forEach { f ->
                        @Suppress("UNCHECKED_CAST")
                        val raw = mapper.readValue(f.toFile(), Map::class.java) as Map<String, Any?>
                        val cases = raw["cases"] as? List<*>
                        if (!cases.isNullOrEmpty() && raw["format"] != null) suites.add(raw)
                    }
            }
        }
    }
    check(suites.isNotEmpty()) { "no compliance suites under $root" }
    return suites
}

private fun inputBytes(c: Map<String, Any?>): ByteArray {
    val enc = c["input_encoding"]?.toString() ?: "utf-8"
    val input = c["input"]?.toString() ?: ""
    if (enc == "hex") {
        val compact = input.replace(Regex("\\s+"), "")
        return if (compact.isEmpty()) ByteArray(0) else compact.chunked(2).map { it.toInt(16).toByte() }.toByteArray()
    }
    return input.toByteArray()
}

private fun runOne(suite: Map<String, Any?>, c: Map<String, Any?>, a: Adapter): MutableMap<String, Any?> {
    val base = mutableMapOf<String, Any?>(
        "id" to c["id"],
        "language" to "kotlin",
        "serializer" to a.name,
        "serializer_version" to a.version,
        "format" to suite["format"],
        "standard" to suite["standard"],
        "standard_url" to suite["standard_url"],
        "version" to suite["version"],
        "version_key" to "${suite["format"]}.${suite["version"]}",
        "requirement" to c["requirement"],
        "expect" to c["expect"],
        "section" to c["section"],
        "section_title" to c["section_title"],
        "section_url" to c["section_url"],
        "paragraph" to c["paragraph"],
        "title" to c["title"],
        "input" to c["input"],
        "input_encoding" to (c["input_encoding"] ?: "utf-8"),
        "detail" to "",
        "observed" to "",
        "outcome" to "pass",
    )
    var got: Any? = null
    var ok = true
    var err = ""
    try {
        got = a.decode(inputBytes(c), c["schema"]?.toString().orEmpty())
    } catch (ex: Exception) {
        ok = false
        err = "${ex.javaClass.simpleName}: ${ex.message}"
    }
    when (c["expect"]) {
        "any" -> base["observed"] = if (ok) preview(got) else err
        "reject" -> if (ok) {
            base["outcome"] = "fail"
            base["detail"] = "parser accepted input the spec requires to be rejected"
            base["observed"] = "accepted as ${preview(got)}"
        } else {
            base["observed"] = err
        }
        else -> if (!ok) {
            base["outcome"] = "fail"
            base["detail"] = "parser rejected input the spec requires to accept"
            base["observed"] = err
        } else {
            base["observed"] = preview(got)
        }
    }
    return base
}

private fun preview(v: Any?): String {
    val s = try {
        jacksonObjectMapper().writeValueAsString(v)
    } catch (_: Exception) {
        v.toString()
    }
    return if (s.length > 120) s.substring(0, 117) + "..." else s
}

private fun printSummary(results: List<Map<String, Any?>>, adapterErrs: List<String>) {
    var p = 0; var f = 0; var s = 0; var e = 0
    val by = sortedMapOf<String, IntArray>()
    for (r in results) {
        when (r["outcome"]) {
            "pass" -> p++
            "fail" -> f++
            "skip" -> s++
            else -> e++
        }
        val k = "${r["standard"]} (${r["version"]}) × ${r["serializer"]}"
        val c = by.getOrPut(k) { IntArray(3) }
        when (r["outcome"]) {
            "pass" -> c[0]++
            "fail" -> c[1]++
            else -> c[2]++
        }
    }
    println("Serialization compliance (library deviations are catalogued, not a red build)")
    println("  $p pass  $f fail  $s skip  $e error  ${results.size} total")
    if (adapterErrs.isNotEmpty()) {
        println("  Adapter errors:")
        adapterErrs.forEach { println("    - $it") }
    }
    println("  By suite × adapter:")
    by.forEach { (k, c) -> println("    $k: ${c[0]} pass, ${c[1]} fail, ${c[2]} skip") }
}

private fun writeReport(path: Path, results: List<Map<String, Any?>>, adapterErrs: List<String>) {
    val p = results.count { it["outcome"] == "pass" }
    val f = results.count { it["outcome"] == "fail" }
    val s = results.count { it["outcome"] == "skip" }
    val e = results.size - p - f - s
    val fmts = results.map { it["format"].toString() }.distinct()
    val doc = mapOf(
        "schema" to "gld.dashboard.compliance/1",
        "generated_at" to Instant.now().toString(),
        "language" to "kotlin",
        "languages" to listOf("kotlin"),
        "policy" to "report-only",
        "scope" to mapOf("formats" to fmts),
        "passed" to p,
        "failed" to f,
        "skipped" to s,
        "errors" to e,
        "catalog_errors" to emptyList<String>(),
        "serializer_errors" to adapterErrs,
        "results" to results,
    )
    Files.createDirectories(path.parent)
    path.writeText(jacksonObjectMapper().writerWithDefaultPrettyPrinter().writeValueAsString(doc) + "\n")
}
